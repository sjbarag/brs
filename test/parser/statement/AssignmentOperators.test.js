const brs = require("brs");
const { Lexeme } = brs.lexer;
const { BrsString, Int32 } = brs.types;

const { token, identifier, EOF } = require("../ParserTests");

describe("parser assignment operators", () => {
    let parser;

    beforeEach(() => {
        parser = new brs.parser.Parser();
    });


    test("+=", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.PlusEqual),
            token(Lexeme.String, `"lorem"`, new BrsString("lorem")),
            EOF
        ]);

        expect(errors).toEqual([]);
        expect(statements).not.toBeFalsy();
        expect(statements).toMatchSnapshot();
    });

    test("-=", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.MinusEqual),
            token(Lexeme.Integer, "1", new Int32(1)),
            EOF
        ]);

        expect(errors).toEqual([]);
        expect(statements).not.toBeFalsy();
        expect(statements).toMatchSnapshot();
    });

    test("^=", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.CaretEqual),
            token(Lexeme.Integer, "2", new Int32(2)),
            EOF
        ]);

        expect(errors).toEqual([]);
        expect(statements).not.toBeFalsy();
        expect(statements).toMatchSnapshot();
    });

    test("*=", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.StarEqual),
            token(Lexeme.Integer, "3", new Int32(3)),
            EOF
        ]);

        expect(errors).toEqual([]);
        expect(statements).not.toBeFalsy();
        expect(statements).toMatchSnapshot();
    });

    test("/=", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.SlashEqual),
            token(Lexeme.Integer, "4", new Int32(4)),
            EOF
        ]);

        expect(errors).toEqual([]);
        expect(statements).not.toBeFalsy();
        expect(statements).toMatchSnapshot();
    });

    test("\\=", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.BackslashEqual),
            token(Lexeme.Integer, "5", new Int32(5)),
            EOF
        ]);

        expect(errors).toEqual([]);
        expect(statements).not.toBeFalsy();
        expect(statements).toMatchSnapshot();
    });

    test("<<=", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.LeftShiftEqual),
            token(Lexeme.Integer, "6", new Int32(6)),
            EOF
        ]);

        expect(errors).toEqual([]);
        expect(statements).not.toBeFalsy();
        expect(statements).toMatchSnapshot();
    });

    test(">>=", () => {
        let { statements, errors } = parser.parse([
            identifier("_"),
            token(Lexeme.RightShiftEqual),
            token(Lexeme.Integer, "7", new Int32(7)),
            EOF
        ]);

        expect(errors).toEqual([]);
        expect(statements).not.toBeFalsy();
        expect(statements).toMatchSnapshot();
    });
});