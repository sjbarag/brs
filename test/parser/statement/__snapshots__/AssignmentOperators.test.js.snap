// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser assignment operators *= 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "StarEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
    },
    "type": "Assignment",
    "value": Binary {
      "left": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "_",
        },
        "type": "Variable",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 3,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "StarEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser assignment operators += 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "PlusEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
    },
    "type": "Assignment",
    "value": Binary {
      "left": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "_",
        },
        "type": "Variable",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": BrsString {
          "kind": 3,
          "value": "lorem",
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "PlusEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser assignment operators /= 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "SlashEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
    },
    "type": "Assignment",
    "value": Binary {
      "left": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "_",
        },
        "type": "Variable",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 4,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "SlashEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser assignment operators <<= 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "LeftShiftEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
    },
    "type": "Assignment",
    "value": Binary {
      "left": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "_",
        },
        "type": "Variable",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 6,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "LeftShiftEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser assignment operators >>= 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "RightShiftEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
    },
    "type": "Assignment",
    "value": Binary {
      "left": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "_",
        },
        "type": "Variable",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 7,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "RightShiftEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser assignment operators \\= 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "BackslashEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
    },
    "type": "Assignment",
    "value": Binary {
      "left": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "_",
        },
        "type": "Variable",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 5,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "BackslashEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser assignment operators -= 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "MinusEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
    },
    "type": "Assignment",
    "value": Binary {
      "left": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "_",
        },
        "type": "Variable",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 1,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "MinusEqual",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": undefined,
      },
      "type": "Binary",
    },
  },
]
`;
